pkg/updatestatus/updatestatuscontroller.go

```diff
@@ -0,0 +1,234 @@
+// the insight and the insight itself, serialized as YAML. Passing serialized avoids shared data access problems. Until
+// we have the Status API we need to serialize ourselves anyway.
+type informerMsg struct {
+	uid     string
+	insight []byte
+}
+
+type sendInsightFn func(insight informerMsg)
+
+func isStatusInsightKey(k string) bool {
+	return strings.HasPrefix(k, "usc-")
+}
+
+// updateStatusController is a controller that collects insights from informers and maintains a ConfigMap with the insights
+// until we have a proper UpdateStatus API. The controller maintains an internal desired content of the ConfigMap (even
+// if it does not exist in the cluster) and updates it in the cluster when new insights are received, or when the ConfigMap
+// changes in the cluster. The controller only maintains the ConfigMap in the cluster if it exists, it does not create it
+// itself (this serves as a simple opt-in mechanism).
+//
+// The communication between informers (insight producers) and this controller is performed via a channel.
+// The newUpdateStatusController constructor calls the startCollectingInsights function that creates the channel spawns
+// a goroutine that consumes the channel and updates the internal ConfigMap state with the received insights. The constructor
+// then creates a function that other controllers can use to send insights to this controller, using the informerMsg
+// structure as a data transfer object.
+type updateStatusController struct {
+	configMaps corev1client.ConfigMapInterface
+
+	// statusApi is the desired state of the status API ConfigMap. It is updated when new insights are received.
+	// Any access to the struct should be done with the lock held.
+	statusApi struct {
+		sync.Mutex
+		cm *corev1.ConfigMap
+
+		// processed is the number of insights processed, used for testing
+		processed int
+	}
+
+	recorder events.Recorder
+}
+
+// newUpdateStatusController creates a new update status controller and returns it. The second return value is a function
+// the other controllers should use to send insights to this controller.
+func newUpdateStatusController(
+	ctx context.Context,
+	coreClient kubeclient.Interface,
+	coreInformers kubeinformers.SharedInformerFactory,
+	recorder events.Recorder,
+) (factory.Controller, sendInsightFn) {
+	uscRecorder := recorder.WithComponentSuffix("update-status-controller")
+
+	c := &updateStatusController{
+		configMaps: coreClient.CoreV1().ConfigMaps(uscNamespace),
+		recorder:   uscRecorder,
+	}
+
+	cmInformer := coreInformers.Core().V1().ConfigMaps().Informer()
+	controller := factory.New().
+		// call sync every 5 minutes or on CM events in the openshift-cluster-version namespace
+		WithSync(c.sync).ResyncEvery(5*time.Minute).
+		WithFilteredEventsInformersQueueKeysFunc(cmNameKey, nsFilter(uscNamespace), cmInformer).
+		ToController("UpdateStatusController", c.recorder)
+
+	// Spawn the insight collecting goroutine
+	sendInsight := c.startCollectingInsights(ctx)
```

Comment:
I'm a bit squeamish about launching a goroutine from the `new*` helper, instead of waiting on a `Run` or `Sync` call in the returned `factory.Controller`.  But I'm also not sure how to restructure this to avoid that yet...

Reply:
/hold 

will address

Reply:
Reworked in 25061544f3fcd1aa2ad04e50ecd98c2944a9e414 to use `WithPostStartHooks`, this really improved the code structure IMO so thanks for this comment! Besides being cleaner, it also partially addressed your [other comment](https://github.com/openshift/cluster-version-operator/pull/1091#discussion_r1810683168) about exhaustively processing all messages. Launching the receiver in the posthook has the advantage of having access to the controller queue, so the receiver can queue the CM for standard controller sync. The controller queue afaik deduplicates items which means if we receive multiple messages between syncs we will still process all of them and change the controller's internal state for each, but will only sync once because the queued syncs will be deduplicated.